### PR TITLE
bpo-31815: Make itertools iterators interruptible with Ctrl-C.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue31815

### CODE REVIEW

File: Modules/itertoolsmodule.c [hunk 1]
```diff
@@ -948,6 +953,9 @@ cycle_next(cycleobject *lz)
     }
     if (PyList_GET_SIZE(lz->saved) == 0)
         return NULL;
+    if ((lz->index & 0xfff) == 0 && PyErr_CheckSignals()) {
```
Comment (vstinner) [inline Modules/itertoolsmodule.c]: Please don't hardcode this constant but put a constant at the top of the file.

**Comment by ncoghlan** [comment]
Interesting - I forgot that the check to make `PyErr_CheckSignals()` a no-op outside the main thread was an internal check inside the function, rather than a rule we applied externally.

Given that, +1 from me.

**Review by ncoghlan — APPROVED** [review]

**Comment by jdemeyer** [comment]
I know that this is probably not going to merged, but let me comment anyway: I don't like the fact that you're calling `PyErr_CheckSignals()` only periodically instead of at every iteration. This introduces unneeded complexity, such as the added `sigcount` struct member in some cases.

I guess that this was done for efficiency. However, if we make `PyErr_CheckSignals` more efficient (moving the `is_tripped` check to a macro), then calling `PyErr_CheckSignals()` every time might actually be more efficient than keeping track of the iteration count and checking `(sigcount & 0xffff) == 0`.

**Comment by vstinner** [comment]
This PR should be closed, since the related bpo was closed https://bugs.python.org/issue31815#msg320356 no?

**Comment by serhiy-storchaka** [comment]
@jdemeyer, if you can make `PyErr_CheckSignals` so efficient the overhead will become negligible, this issue can be revived.

**Comment by jdemeyer** [comment]
> this issue can be revived.

As far as I can see from the bpo-31815 discussion, efficiency was not the reason for rejection of this PR.

**Comment by jdemeyer** [comment]
> if you can make `PyErr_CheckSignals` so efficient the overhead will become negligible

Of course I can. As usual, the hard part in CPython is not the coding, but convincing a core dev to merge your code. So I'm only going to bother if it's clear that this is a wanted feature.

**Pull request closed without merging**